File: src/agent_sdk/logging_.py
from __future__ import annotations

import json
import threading
from datetime import datetime, timezone
from pathlib import Path
from typing import Any
from uuid import UUID


# ##################################################################
# default log base
# fallback directory used when config does not specify a log location
_DEFAULT_LOG_BASE = Path.home() / ".agent-sdk" / "logs"


# ##################################################################
# conversation logger
# writes structured logs for a single conversation into its own directory
# thread-safe: all file writes are protected by a lock
# all io errors are swallowed — logging failures must never crash the caller
class ConversationLogger:

    # ##################################################################
    # init
    # creates the conversation directory and writes initial meta.json
    # log_base defaults to ~/.agent-sdk/logs when not provided
    def __init__(
        self,
        conversation_uuid: UUID,
        *,
        name: str | None = None,
        tier: str | None = None,
        provider: str | None = None,
        model: str | None = None,
        log_base: Path | str | None = None,
    ) -> None:
        self._uuid = conversation_uuid
        self._lock = threading.Lock()
        self._closed = False

        base = Path(log_base) if log_base is not None else _DEFAULT_LOG_BASE
        self._log_dir = base / str(conversation_uuid)

        try:
            self._log_dir.mkdir(parents=True, exist_ok=True)
            meta = {
                "conversation_id": str(conversation_uuid),
                "name": name,
                "tier": tier,
                "provider": provider,
                "model": model,
                "created_at": _now_iso(),
            }
            (self._log_dir / "meta.json").write_text(json.dumps(meta, indent=2), encoding="utf-8")
        except OSError:
            pass

    # ##################################################################
    # log dir property
    # returns the path to this conversation's log directory
    @property
    def log_dir(self) -> Path:
        return self._log_dir

    # ##################################################################
    # log event
    # appends a single json line to events.jsonl
    # timestamp is added automatically — callers supply event_type and kwargs
    def log_event(self, event_type: str, **kwargs: Any) -> None:
        entry: dict[str, Any] = {"ts": _now_iso(), "event": event_type}
        entry.update(kwargs)
        self._write_jsonl(entry)

    # ##################################################################
    # close
    # writes a final conversation_end event and marks the logger closed
    def close(self, **kwargs: Any) -> None:
        if not self._closed:
            self._closed = True
            self.log_event("conversation_end", **kwargs)

    # ##################################################################
    # write jsonl
    # internal: appends a dict as one json line to events.jsonl
    # acquires the lock and silently ignores all io errors
    def _write_jsonl(self, entry: dict[str, Any]) -> None:
        try:
            line = json.dumps(entry, default=str) + "\n"
            with self._lock:
                with open(self._log_dir / "events.jsonl", "a", encoding="utf-8") as fh:
                    fh.write(line)
        except OSError:
            pass


# ##################################################################
# now iso
# returns the current utc time as an iso 8601 string
def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()

File: go/logging.go
package dazagentsdk

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
)

// defaultLogBase returns ~/.daz-agent-sdk/logs as the fallback log directory.
func defaultLogBase() string {
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	return filepath.Join(home, ".daz-agent-sdk", "logs")
}

// ConversationLogger writes structured JSONL logs for a single conversation
// into its own directory. All file writes are protected by a mutex.
// IO errors are swallowed — logging failures must never crash the caller.
type ConversationLogger struct {
	conversationID uuid.UUID
	name           string
	logDir         string
	mu             sync.Mutex
	closed         bool
}

// NewConversationLogger creates a new logger, creates the log directory, and
// writes the initial meta.json. If no conversationID is provided, a new UUID
// is generated. The log directory is determined from cfg.Logging.Directory,
// falling back to ~/.daz-agent-sdk/logs.
func NewConversationLogger(name string, cfg *Config, conversationID ...uuid.UUID) *ConversationLogger {
	var id uuid.UUID
	if len(conversationID) > 0 {
		id = conversationID[0]
	} else {
		id = uuid.New()
	}

	base := defaultLogBase()
	if cfg != nil && cfg.Logging.Directory != "" {
		dir := cfg.Logging.Directory
		// Expand ~ prefix
		if strings.HasPrefix(dir, "~/") {
			if home, err := os.UserHomeDir(); err == nil {
				dir = filepath.Join(home, dir[2:])
			}
		}
		base = dir
	}

	logDir := filepath.Join(base, id.String())

	cl := &ConversationLogger{
		conversationID: id,
		name:           name,
		logDir:         logDir,
	}

	// Create directory and write meta.json — swallow errors
	if err := os.MkdirAll(logDir, 0o755); err == nil {
		meta := map[string]any{
			"conversation_id": id.String(),
			"name":            name,
			"created_at":      nowISO(),
		}
		if data, err := json.MarshalIndent(meta, "", "  "); err == nil {
			_ = os.WriteFile(filepath.Join(logDir, "meta.json"), data, 0o644)
		}
	}

	return cl
}

// LogDir returns the path to this conversation's log directory.
func (cl *ConversationLogger) LogDir() string {
	return cl.logDir
}

// ConversationID returns the UUID for this conversation.
func (cl *ConversationLogger) ConversationID() uuid.UUID {
	return cl.conversationID
}

// LogEvent appends a single JSON line to events.jsonl.
// The timestamp is added automatically. Extra key-value pairs from the
// optional map are merged into the event.
func (cl *ConversationLogger) LogEvent(eventType string, extra ...map[string]any) {
	entry := map[string]any{
		"ts":    nowISO(),
		"event": eventType,
	}
	for _, m := range extra {
		for k, v := range m {
			entry[k] = v
		}
	}
	cl.writeJSONL(entry)
}

// Close writes a final conversation_end event and marks the logger closed.
// Subsequent calls to Close are no-ops.
func (cl *ConversationLogger) Close(extra ...map[string]any) {
	cl.mu.Lock()
	if cl.closed {
		cl.mu.Unlock()
		return
	}
	cl.closed = true
	cl.mu.Unlock()

	cl.LogEvent("conversation_end", extra...)
}

// writeJSONL appends a dict as one JSON line to events.jsonl.
// Acquires the lock and silently ignores all IO errors.
func (cl *ConversationLogger) writeJSONL(entry map[string]any) {
	data, err := json.Marshal(entry)
	if err != nil {
		return
	}
	line := append(data, '\n')

	cl.mu.Lock()
	defer cl.mu.Unlock()

	f, err := os.OpenFile(filepath.Join(cl.logDir, "events.jsonl"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return
	}
	defer f.Close()
	_, _ = f.Write(line)
}

// nowISO returns the current UTC time as an ISO 8601 string.
func nowISO() string {
	return time.Now().UTC().Format(time.RFC3339Nano)
}
